mitodl/ocw-studio
# Cc/link parser and resolveuid

#### Pre-Flight checklist


#### What are the relevant tickets https://github.com/mitodl/ocw-studio/issues/1116

#### What's this PR do?
This PR adds two new `markdown_cleanup` rules:
- `resolveuid`: converts markdown links/images with `resolveuid` to shortcodes or rootrelative links, depending on whether they are cross-site or not.
- `link_logging`: Logs all markdown links within a course to a CSV 

Most of the new logic is in `link_parser.py` which defines a Pyparsing-based markdown link parser. See comments that.

#### How should this be manually tested?

1. Run `docker-compose run --rm web python manage.py markdown_cleanup resolveuid -o resolveuid.csv` to generate a list of changes that will be made by the resolve uid rule.

    -  The resulting CSV will include some rows that have not changed, but it is not too big, only ~450 rows
    - the command takes about 3-5 minutes to run.
2. Run the above with `--commit` and check that the appropriate changes were made if you like. 

It might also be fun to run `docker-compose run --rm web python manage.py markdown_cleanup link_logging -o link_logging.csv` to log all markdown links. This takes about 45 minutes. 

#### Where should the reviewer start?
Suggest starting with `link_parser.py`.
